src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,170 @@
+             )
+
+  open Bimodule bimodule
+    renaming (Carrierᴹ to M
+             )
+
+  open AbelianGroup +ᴹ-abelianGroup
+    renaming (setoid to setoidᴹ
+             ; isEquivalence to isEquivalenceᴹ
+             )
+
+  open IsEquivalence isEquivalenceᴹ
+    renaming (sym to symᴹ)
+
+  +ᴺ-bimodule = DirectProduct.bimodule TensorUnit.bimodule bimodule
+
+  open Bimodule +ᴺ-bimodule public using ()
+    renaming (Carrierᴹ to N
+             ; _≈ᴹ_ to _≈ᴺ_
+             ; _+ᴹ_ to _+ᴺ_
+             ; 0ᴹ to 0ᴺ
+             ; -ᴹ_ to -ᴺ_
+             ; +ᴹ-abelianGroup to +ᴺ-abelianGroup
+             )
+
+  ιᴿ : R → N
+  ιᴿ r = r , 0ᴹ
+
+  ιᴹ : M → N
+  ιᴹ m = 0ᴿ , m
+
+  1ᴺ : N
+  1ᴺ = ιᴿ 1ᴿ
+
+  open AbelianGroup +ᴺ-abelianGroup
+    renaming (isAbelianGroup to +ᴺ-isAbelianGroup)
+
+  open Definitions _≈ᴺ_
+
+  infixl 7 _*ᴺ_
+
+  _*ᴺ_ : Op₂ N
+  (r₁ , m₁) *ᴺ (r₂ , m₂) = r₁ *ᴿ r₂ , r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂
+
+  *ᴺ-identity : Identity 1ᴺ _*ᴺ_
+  *ᴺ-identity = lᴺ , rᴺ
+    where
+      open ≈-Reasoning setoidᴹ
+      lᴺ : LeftIdentity 1ᴺ _*ᴺ_
+      lᴺ (r , m) = (*ᴿ-identityˡ r) , (begin
+        (1ᴿ *ₗ m +ᴹ 0ᴹ *ᵣ r) ≈⟨ +ᴹ-cong (*ₗ-identityˡ m) (*ᵣ-zeroˡ r) ⟩
+        (m +ᴹ 0ᴹ) ≈⟨ +ᴹ-identityʳ m ⟩
+        m ∎)
+      rᴺ : RightIdentity 1ᴺ _*ᴺ_
+      rᴺ (r , m) = (*ᴿ-identityʳ r) , (begin
+        r *ₗ 0ᴹ +ᴹ m *ᵣ 1ᴿ ≈⟨ +ᴹ-cong (*ₗ-zeroʳ r) (*ᵣ-identityʳ m) ⟩
+        0ᴹ +ᴹ m ≈⟨ +ᴹ-identityˡ m ⟩
+        m ∎)
+
+  *ᴺ-cong : Congruent₂ _*ᴺ_
+  *ᴺ-cong (r₁ , m₁) (r₂ , m₂) = *ᴿ-cong r₁ r₂ , +ᴹ-cong (*ₗ-cong r₁ m₂) (*ᵣ-cong m₁ r₂)
+
+  *ᴺ-assoc : Associative _*ᴺ_
+  *ᴺ-assoc (r₁ , m₁) (r₂ , m₂) (r₃ , m₃) = *ᴿ-assoc r₁ r₂ r₃ , (begin
+    (r₁ *ᴿ r₂) *ₗ m₃ +ᴹ (r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂) *ᵣ r₃
+      ≈⟨ +ᴹ-cong (*ₗ-assoc r₁ r₂ m₃) (*ᵣ-distribʳ r₃ (r₁ *ₗ m₂) (m₁ *ᵣ r₂)) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃) +ᴹ ((r₁ *ₗ m₂) *ᵣ r₃ +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃)
+       ≈⟨ +ᴹ-congˡ (+ᴹ-congʳ (*ₗ-*ᵣ-assoc r₁ m₂ r₃)) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃) +ᴹ (r₁ *ₗ (m₂ *ᵣ r₃) +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃)
+      ≈⟨ symᴹ (+ᴹ-assoc (r₁ *ₗ (r₂ *ₗ m₃)) (r₁ *ₗ (m₂ *ᵣ r₃)) ((m₁ *ᵣ r₂) *ᵣ r₃)) ⟩
+    (r₁ *ₗ (r₂ *ₗ m₃) +ᴹ r₁ *ₗ (m₂ *ᵣ r₃)) +ᴹ (m₁ *ᵣ r₂) *ᵣ r₃
+      ≈⟨ +ᴹ-cong (symᴹ (*ₗ-distribˡ r₁ (r₂ *ₗ m₃) (m₂ *ᵣ r₃))) (*ᵣ-assoc m₁ r₂ r₃) ⟩
+    r₁ *ₗ (r₂ *ₗ m₃ +ᴹ m₂ *ᵣ r₃) +ᴹ m₁ *ᵣ (r₂ *ᴿ r₃) ∎)
+    where
+      open ≈-Reasoning setoidᴹ
+
+  *ᴺ-distrib-+ᴺ : _*ᴺ_ DistributesOver _+ᴺ_
+  *ᴺ-distrib-+ᴺ = lᴺ , rᴺ
+    where
+      open ≈-Reasoning setoidᴹ
+      open Consequences setoidᴹ
+      +ᴹ-middleFour = comm∧assoc⇒middleFour +ᴹ-cong +ᴹ-comm +ᴹ-assoc
+      lᴺ : _*ᴺ_ DistributesOverˡ _+ᴺ_
+      lᴺ (r₁ , m₁) (r₂ , m₂) (r₃ , m₃) = *ᴿ-distribˡ-+ᴿ r₁ r₂ r₃ , (begin
+        r₁ *ₗ (m₂ +ᴹ m₃) +ᴹ m₁ *ᵣ (r₂ +ᴿ r₃)
+          ≈⟨ +ᴹ-cong (*ₗ-distribˡ r₁ m₂ m₃) (*ᵣ-distribˡ m₁ r₂ r₃) ⟩
+        (r₁ *ₗ m₂ +ᴹ r₁ *ₗ m₃) +ᴹ (m₁ *ᵣ r₂ +ᴹ m₁ *ᵣ r₃)
+          ≈⟨ +ᴹ-middleFour (r₁ *ₗ m₂) (r₁ *ₗ m₃) (m₁ *ᵣ r₂) (m₁ *ᵣ r₃) ⟩
+        (r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂) +ᴹ (r₁ *ₗ m₃ +ᴹ m₁ *ᵣ r₃) ∎)
+      rᴺ : _*ᴺ_ DistributesOverʳ _+ᴺ_
```

Comment:
given how much you like extra white space (way more than I do!), I am rather surprise to see none here. The definitions here do blur into each other, so I would have appreciated the use of space to make that clearer).

Have you considered using `a,b,c,d` and `m,n,o,p` instead of subscripts?  If I were doing this, I'd be tempted to use font to distinguish between Ring and Module variables (i.e. bold for Module).

Reply:
Re: whitespace, yes, on reflection (this development is now maybe 18months old, so needs considerable updating!) will do so!
Re: variables. Yes, this is perhaps more annoying, and we don't (yet) have conventions for these things. I agree subscripts can be awful, and I hadn't thought of using font/weight distinctions for this... will consider further.

Reply:
On further thoughts, how do I get boldface in the Unicode/Emacs-agda2-mode input encoding? 
I'd like to try it, but don't quite understand how!?

Reply:
You can't quite get 'bold' per se, but for m, you can try `\Mim`, `\Mcm`, `\MCm`, `\MIm`, `Mfm` to get a variety of weights and styles.  Here `\MIm` appeals most to my sense of style.

Reply:
Thanks for the hints! I'll try this on other PRs I think, and not try to fiddle *too* much more with this one!